### ISSUE TITLE
Maven docker images are using shell entrypoints

### ISSUE THREAD

**Issue by jroper** [lightbend/openshift-deployment-guide#80]
Our docker images built by our Maven instructions are using shell entrypoints, ie, we supply a command that is passed to a shell to be executed. The consequence of this is that when the container is terminated, the SIGTERM gets swallowed by that shell, shutdown times out, and our java process gets killed - this is particularly nefarious for Akka clusters since it means nodes never leave the cluster gracefully and need to be downed instead.

One solution is to use an exec entry point, documented here:

https://dmp.fabric8.io/#misc-startup

However, I don't think it's that straight forward, because currently we pass the `$JAVA_OPTS` environment variable as a command argument, and I doubt environment variables will work when doing an exec. Even if they did, we set multiple arguments in the environment variable, but the contents of that variable would be treated by one argument as the java command.

All that said, I think there is a simple solution, simply prepend `exec` to the command, and then java process will replace the shell process (because it will use the bash built in `exec` command which execs without forking first). So, the config becomes:

```xml
                    <entryPoint>
                       exec java $JAVA_OPTS -cp '/maven/*' play.core.server.ProdServerStart
                    </entryPoint> 
```

**Comment by TimMoore** [comment]
Yes, good point. However, running `java` as PID 1 [has its own drawbacks](https://github.com/lightbend/sbt-reactive-app/issues/118). Still, this is an incremental improvement that solves the more common problem, so it's worth doing in the short term.

Longer term, we should try using [`tini`](https://github.com/krallin/tini).

**Comment by ihostage** [comment]
@TimMoore @jroper 
Hi, folks! 👋 
I right understand, that using `run-java.sh` has the same problem with gracefull shutdown Akka Cluster?
```xml
<plugin>
  <groupId>io.fabric8</groupId>
  <artifactId>docker-maven-plugin</artifactId>
  <version>0.30.0</version>
  <dependencies>
    <dependency>
      <groupId>io.fabric8</groupId>
      <artifactId>run-java-sh</artifactId>
      <version>1.3.4</version>
    </dependency>
  </dependencies>
...
    <cmd>/opt/bin/run-java.sh</cmd>
...
</plugin>
```

**Comment by jroper** [comment]
`run-java-exec` execs:

https://github.com/fabric8io-images/run-java-sh/blob/master/fish-pepper/run-java-sh/fp-files/run-java.sh#L605

So, as you use an exec rather than shell entrypoint, your process will end up with PID 1 and it will correctly handle SIGTERM, SIGINT etc.

**Comment by ihostage** [comment]
Thank you, James @jroper! 
You are right, the process has a PID 1
```
# ps faux
PID   USER     TIME  COMMAND
    1 root     25:13 java -Dconfig.resource=application-docker.conf -Dlogger.file=/opt/lib/logback-docker.xml -XshowSettings:vm ...
```

Thus `io.fabric8#run-java-sh` can be yet another solution 😄
